idx/bigtable/bigtable.go

```diff
@@ -0,0 +1,544 @@
+						time.Sleep(time.Second)
+					} else {
+						return
+					}
+				}
+			}()
+		}
+	}
+
+	// check if we need to save to bigtable.
+	now := uint32(time.Now().Unix())
+	if archive.LastSave < (now - updateInterval32) {
+		archive = b.updateBigtable(now, inMemory, archive, partition)
+	}
+
+	stat.Value(time.Since(pre))
+	return archive, oldPartition, inMemory
+}
+
+func (b *BigtableIdx) updateBigtable(now uint32, inMemory bool, archive idx.Archive, partition int32) idx.Archive {
+	// if the entry has not been saved for 1.5x updateInterval
+	// then perform a blocking save.
+	if archive.LastSave < (now - updateInterval32 - (updateInterval32 / 2)) {
+		log.Debugf("bigtable-idx: updating def %s in index.", archive.MetricDefinition.Id)
+		b.writeQueue <- writeReq{recvTime: time.Now(), def: &archive.MetricDefinition}
+		archive.LastSave = now
+		b.MemoryIdx.UpdateArchive(archive)
+	} else {
+		// perform a non-blocking write to the writeQueue. If the queue is full, then
+		// this will fail and we wont update the LastSave timestamp. The next time
+		// the metric is seen, the previous lastSave timestamp will still be in place and so
+		// we will try and save again.  This will continue until we are successful or the
+		// lastSave timestamp become more then 1.5 x UpdateInterval, in which case we will
+		// do a blocking write to the queue.
+		select {
+		case b.writeQueue <- writeReq{recvTime: time.Now(), def: &archive.MetricDefinition}:
+			archive.LastSave = now
+			b.MemoryIdx.UpdateArchive(archive)
+		default:
+			statSaveSkipped.Inc()
+			log.Debugf("bigtable-idx: writeQueue is full, update of %s not saved this time", archive.MetricDefinition.Id)
+		}
+	}
+
+	return archive
+}
+
+func (b *BigtableIdx) rebuildIndex() {
+	log.Info("bigtable-idx: Rebuilding Memory Index from metricDefinitions in bigtable")
+	pre := time.Now()
+
+	var staleTs uint32
+	if maxStale != 0 {
+		staleTs = uint32(time.Now().Add(maxStale * -1).Unix())
+	}
+	num := 0
+	var defs []schema.MetricDefinition
+	for _, partition := range cluster.Manager.GetPartitions() {
+		defs = b.LoadPartition(partition, defs[:0], staleTs)
+		num += b.MemoryIdx.Load(defs)
+	}
+
+	log.Infof("bigtable-idx: Rebuilding Memory Index Complete. Imported %d. Took %s", num, time.Since(pre))
+}
+
+func (b *BigtableIdx) LoadPartition(partition int32, defs []schema.MetricDefinition, cutoff uint32) []schema.MetricDefinition {
+	ctx := context.Background()
+	rr := bigtable.PrefixRange(fmt.Sprintf("%d_", partition))
+	defsBySeries := make(map[string][]schema.MetricDefinition)
+	err := b.tbl.ReadRows(ctx, rr, func(r bigtable.Row) bool {
+		def := schema.MetricDefinition{}
+		err := RowToSchema(r, &def)
+		if err != nil {
+			log.Errorf("bigtable-idx: failed to marshal row to metricDef. %s", err)
+			return false
+		}
+		log.Debugf("bigtable-idx: found def %+v", def)
+		defsBySeries[def.Name] = append(defsBySeries[def.Name], def)
+		return true
+	}, bigtable.RowFilter(bigtable.FamilyFilter(COLUMN_FAMILY)))
```

Comment:
the column family filter is efficient right? it'll just direct reads at the correct CF, and not just scan all data and drop everything it doesn't need ?

Reply:
filters are applied to rows after the rows are read.  We only use 1 column Family (which is why it is a constant).  So will have 0 impact on performance.

Reply:
but the store also uses a bunch of CF's ?

Reply:
The Bigtable schema's were designed in collaboration with Google's Bigtable team.

Reply:
i realize that. my confusion was because I thought 1 CF had multiple tables, but it's the other way around. and idx and store use different tables, so that's why there's only 1 CF for the idx.